pkg/auth/spire/delegate.go

```diff
@@ -0,0 +1,286 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package spire
+
+import (
+	"bytes"
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	delegatedidentityv1 "github.com/spiffe/spire-api-sdk/proto/spire/api/agent/delegatedidentity/v1"
+	spiffeTypes "github.com/spiffe/spire-api-sdk/proto/spire/api/types"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/pflag"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/cilium/cilium/pkg/auth/certs"
+	"github.com/cilium/cilium/pkg/backoff"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+type SpireDelegateClient struct {
+	cfg SpireDelegateConfig
+	log logrus.FieldLogger
+
+	connectionAttempts int
+
+	stream      delegatedidentityv1.DelegatedIdentity_SubscribeToX509SVIDsClient
+	trustStream delegatedidentityv1.DelegatedIdentity_SubscribeToX509BundlesClient
+
+	svidStore   map[string]*delegatedidentityv1.X509SVIDWithKey
+	trustBundle *x509.CertPool
+
+	cancelListenForUpdates context.CancelFunc
+}
+
+type SpireDelegateConfig struct {
+	SpireAdminSocketPath string `mapstructure:"spire-admin-socket-path"`
+	SpiffeTrustDomain    string `mapstructure:"spiffe-trust-domain"`
+}
+
+type certificateProviderResult struct {
+	cell.Out
+
+	CertificateProvider certs.CertificateProvider `group:"certificateProviders"`
+}
+
+var Cell = cell.Module(
+	"spire-delegate",
+	"Spire Delegate API Client",
+	cell.Provide(newSpireDelegateClient),
+	cell.Config(SpireDelegateConfig{}),
+)
+
+func newSpireDelegateClient(lc hive.Lifecycle, cfg SpireDelegateConfig, log logrus.FieldLogger) certificateProviderResult {
+	client := &SpireDelegateClient{
+		cfg:       cfg,
+		log:       log.WithField(logfields.LogSubsys, "spire-delegate"),
+		svidStore: map[string]*delegatedidentityv1.X509SVIDWithKey{},
+	}
+
+	lc.Append(hive.Hook{OnStart: client.onStart, OnStop: client.onStop})
+
+	return certificateProviderResult{
+		CertificateProvider: client,
+	}
+}
+
+func (cfg SpireDelegateConfig) Flags(flags *pflag.FlagSet) {
+	flags.StringVar(&cfg.SpireAdminSocketPath, "spire-admin-socket-path", "/run/spire/sockets/admin.sock", "The path for the SPIRE admin agent Unix socket.")
+	flags.StringVar(&cfg.SpiffeTrustDomain, "spiffe-trust-domain", "spiffe.cilium.io", "The trust domain for the SPIFFE identity.")
```

Comment:
shouldn't we start naming the config properties according to the CFP? https://github.com/cilium/design-cfps/pull/5/files#diff-4e8924edac1a54a8ef7c069b1183ce620bc2de0074704385ae14b76c29b63e14

at least prefixing them with `mesh-auth-`